File: java/com/demod/dcf/networking.java
package com.demod.dcf.networking;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.stub.StreamObserver;
import com.demod.dcf.generated.DCFServiceGrpc;
import com.demod.dcf.generated.DCFMessage;

public class Networking {
    private final ManagedChannel channel;
    private final DCFServiceGrpc.DCFServiceBlockingStub stub;
    private final String mode;

    public Networking(String host, int port, String mode) {
        this.channel = ManagedChannelBuilder.forAddress(host, port).usePlaintext().build();
        this.stub = DCFServiceGrpc.newBlockingStub(channel);
        this.mode = mode;
    }

    public String send(String data, String recipient) {
        DCFMessage request = DCFMessage.newBuilder().setData(data).setRecipient(recipient).build();
        DCFMessage response = stub.sendMessage(request);
        return response.getData();
    }

    // Add stream for receive: stub.receiveStream(Empty.newBuilder().build(), observer);
    public void shutdown() { channel.shutdown(); }
}

File: kotlin/src/main/kotlin/dcf/Networking.kt
// kotlin/src/main/kotlin/dcf/Networking.kt
import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import dcf.generated.DCFMessage
import dcf.generated.DCFServiceGrpcKt

class Networking(
    private val host: String,
    private val port: Int,
    private val mode: Mode = Mode.P2P
) {
    private val channel: ManagedChannel = ManagedChannelBuilder.forAddress(host, port).usePlaintext().build()
    private val stub = DCFServiceGrpcKt.dCFServiceCoroutineStub(channel)

    suspend fun sendMessage(data: String, recipient: String): String {
        val request = DCFMessage.newBuilder().setData(data).setRecipient(recipient).build()
        val response = stub.sendMessage(request)
        return response.data
    }

    fun receiveStream(): Flow<DCFMessage> = flow {
        val response = stub.receiveStream(Unit)  // Assuming Empty request
        response.collect { emit(it) }
    }

    fun close() {
        channel.shutdown()
    }
}
